### PR TITLE
Vpc support

### PR DESCRIPTION
**Pull request by lime-green**
closes issue: https://github.com/aws/chalice/issues/413
cc: @kyleknap @jamesls

### CODE REVIEW

File: chalice/awsclient.py [hunk 1]
```diff
@@ -100,6 +101,14 @@ def get_function_configuration(self, name):
             FunctionName=name)
         return response
 
+    def create_vpc_config(self, security_group_ids, subnet_ids):
+        vpc_config = {}
+        if subnet_ids is not None:
+            vpc_config['SubnetIds'] = subnet_ids
+        if security_group_ids is not None:
```
Comment (achautha) [inline chalice/awsclient.py]: As per [AWS documentation](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-vpcconfig.html), both SubnetIds and SecurityGroupIds are mandatory parameters for VPC configuration.  IMO We should validate if both are present.

File: chalice/policy.py [hunk 2]
```diff
@@ -79,6 +79,8 @@ def generate_policy(self, config):
         app_source = self._osutils.get_file_contents(app_py, binary=False)
         app_policy = policy_from_source_code(app_source)
         app_policy['Statement'].append(CLOUDWATCH_LOGS)
+        if config.subnet_ids or config.security_group_ids:
```
Comment (achautha) [inline chalice/policy.py]: Both parameters should be present for VpcConfig. 'and' condition may be ?

Comment (achautha) [inline chalice/policy.py]: We need to document two new fields in `configfile.rst`

File: chalice/constants.py [hunk 3]
```diff
@@ -80,6 +80,18 @@ def index():
...
 
+VPC_ATTACH_POLICY = {
+    "Effect": "Allow",
+    "Action": [
+        "ec2:CreateNetworkInterface",
+        "ec2:DescribeNetworkInterfaces",
+        "ec2:DetachNetworkInterface",
+        "ec2:DeleteNetworkInterface"
+    ],
+    "Resource": "*"
```
Comment (achautha) [inline chalice/constants.py]: We can use   "Resource": "arn:aws:ec2:::*" .

**Comment by lime-green** [comment]
@achautha thanks for your comments. I'll update this (and actually fix the tests) tomorrow :)

**Comment by lime-green** [comment]
@achautha @kyleknap @jamesls any thoughts?

**Comment by jamesls** [comment]
Hi, thanks for the pull request.  It'd be great to have VPC support in chalice.  This is a fairly large PR so it might take some time to get through it all, but I do have it on my backlog.

File: chalice/deploy/deployer.py [hunk 4]
```diff
@@ -602,17 +602,22 @@ def _deploy_single_lambda_function(self, config, name, handler,
...
+                'environment_variables': config.environment_variables,
+                'runtime': config.lambda_python_version,
+                'handler': handler,
+                'tags': config.tags,
+                'timeout': self._get_lambda_timeout(config),
+                'memory_size': self._get_lambda_memory_size(config),
+            }  # type: Dict[str, Any]
+            if config.security_group_ids is not None:
+                kwargs['security_group_ids'] = config.security_group_ids
+            if config.security_group_ids is not None:
```
Comment (agshew) [inline chalice/deploy/deployer.py]: Should this be checking config.subnet_ids instead of config.security_group_ids again?

**Comment by lstroud** [comment]
Is this still going to be merged? Or, is there another approach that is being favored? 

Is there anything I can do to help hurry this along?

**Comment by ssola** [comment]
Because previous comment didn't had an answer, I will try to ask the same question again. What's the status of this PR? Is there any work left? something that we can help to finish and merge this PR?

Thanks.

**Comment by pkit** [comment]
Probably rebase and fix https://github.com/aws/chalice/pull/457#discussion_r146064645 is needed.
Can confirm that it works pretty well from checkout.

**Comment by lime-green** [comment]
@pkit yes, that is what is needed. I remember it not working with the more specific ARN. Could have been another issue though. I'll see what I can do to get this through, I might need to fork this personally since I don't have write access to this repo any longer.

**Comment by pkit** [comment]
@lime-green I think it indeed does not work with more specific arn, probably because lambda containers are not really ec2? :)

**Comment by swrainey** [comment]
We forked this today and it worked very well. I Would love to see this get into master.

**Comment by lime-green** [comment]
https://github.com/aws/chalice/pull/673 will now replace this PR. Hopefully we can get this through!

**Pull request closed without merging**